src/main/scala/io/iohk/ethereum/vm/Int256.scala

```diff
@@ -0,0 +1,201 @@
+
+  val Modulus: BigInt = BigInt(2).pow(Size * 8)
+
+}
+
+trait Int256Like[T <: Int256Like[T]] extends ScalaNumber with Ordered[T] with ScalaNumericConversions {
+
+  import Int256Like._
+
+  def n: BigInt
+
+  protected def fromBigInt(n: BigInt): T
+
+  protected def fromByteString(n: ByteString): T
+
+  /** Converts a BigInt to a ByteString.
+    * Output ByteString is padded with 0's from the left side up to MaxLength bytes.
+    */
+  lazy val bytes: ByteString = {
+    val bs: ByteString = ByteString(n.toByteArray).takeRight(Size)
+    val padLength: Int = Size - bs.length
+    if (padLength > 0)
+      Zeros.take(padLength) ++ bs
+    else
+      bs
+  }
+
+  def &(that: T): T = fromBigInt(this.n & that.n)
+
+  def |(that: T): T = fromBigInt(this.n | that.n)
+
+  def ^(that: T): T = fromBigInt(this.n ^ that.n)
+
+  def unary_~ : T = fromBigInt(~n)
+
+  def unary_- : T = fromBigInt(-n)
+
+  def +(that: T): T = fromBigInt(this.n + that.n)
+
+  def -(that: T): T = fromBigInt(this.n - that.n)
+
+  def *(that: T): T = fromBigInt(this.n * that.n)
+
+  @throws[ArithmeticException]
+  def div(that: T): T = fromBigInt(this.n / that.n)
+
+  @throws[ArithmeticException]
+  def %(that: T): T = fromBigInt(this.n % that.n)
+
+  @throws[ArithmeticException]
+  def addmod(that: T, modulus: T): T = fromBigInt((this.n + that.n) % modulus.n)
+
+  @throws[ArithmeticException]
+  def mulmod(that: T, modulus: T): T = fromBigInt((this.n * that.n) % modulus.n)
+
+  // https://docs.oracle.com/javase/8/docs/api/java/math/BigInteger.html#modPow-java.math.BigInteger-java.math.BigInteger-
+  // http://stackoverflow.com/questions/8318018/java-how-to-invert-a-biginteger
+  @throws[ArithmeticException]
+  def **(that: T): T = fromBigInt(this.n.modPow(that.n, Modulus))
+
+  def getByte(that: T): T = {
+    if (that.n > Size - 1 || that.n < 0)
+      fromBigInt(BigInt(0))
+    else
+      fromByteString(Zeros.take(Size - 1) :+ this.bytes(that.n.toInt))
+  }
+
+  def compare(that: T): Int = this.n.compare(that.n)
+
+  override def byteValue: Byte = n.byteValue
+
+  override def shortValue: Short = n.shortValue
+
+  override def intValue: Int = n.intValue
+
+  override def longValue: Long = n.longValue
+
+  override def floatValue: Float = n.floatValue
+
+  override def doubleValue: Double = n.doubleValue
+
+  override def isWhole: Boolean = true
+
+  override def underlying: Object = n
+
+}
+
+object UInt256 {
+
+  import Int256Like._
+
+  def apply(n: BigInt): UInt256 = new UInt256(fixBigInt(n))
+
+  def apply(bytes: IndexedSeq[Byte]): UInt256 = {
+    require(bytes.length <= Size, s"Input byte array cannot be longer than $Size: ${bytes.length}")
+    UInt256(bytes.foldLeft(BigInt(0)) { (n, b) => (n << 8) + (b & 0xff) })
+  }
+
+  def apply(int256: Int256): UInt256 = new UInt256(toUnsignedBigInt(int256.n))
+
+  implicit def numericToUInt256[N: Integral](n: N): UInt256 =
+    apply(BigInt(implicitly[Integral[N]].toLong(n)))
```

Comment:
We probably don't want to convert BigInts automatically as that can conversion can be lossy